File: flutter/bicitaxi_conductor/lib/core/services/driver_location_tracker.dart
import 'dart:async';
import 'package:geolocator/geolocator.dart';
import 'request_service.dart';

/// Smart GPS location tracker for conductors.
///
/// Implements intelligent location publishing to reduce Firebase writes:
/// - Samples GPS every 30 seconds locally
/// - Maintains rolling buffer of last 3 positions
/// - Only publishes to Firebase if current position differs >3m from buffer average
class DriverLocationTracker {
  final RequestService _requestService;

  Timer? _locationTimer;
  String? _activeCellId;
  String? _activeRequestId;

  // Rolling buffer of recent positions for smart filtering
  final List<Position> _recentPositions = [];
  static const int _bufferSize = 3;
  static const double _movementThresholdMeters = 3.0;
  static const Duration _samplingInterval = Duration(seconds: 30);

  bool _isTracking = false;

  DriverLocationTracker({required RequestService requestService})
    : _requestService = requestService;

  /// Whether location tracking is currently active.
  bool get isTracking => _isTracking;

  /// Start tracking driver location for a request.
  /// Call this when driver accepts a request.
  void startTracking(String cellId, String requestId) {
    if (_isTracking) {
      stopTracking();
    }

    _activeCellId = cellId;
    _activeRequestId = requestId;
    _recentPositions.clear();
    _isTracking = true;

    print('📍 DriverLocationTracker: Started tracking for request $requestId');

    // Immediately get and publish first position
    _sampleLocation();

    // Start periodic sampling
    _locationTimer = Timer.periodic(_samplingInterval, (_) {
      _sampleLocation();
    });
  }

  /// Stop tracking driver location.
  /// Call this when request is completed or cancelled.
  void stopTracking() {
    _locationTimer?.cancel();
    _locationTimer = null;
    _activeCellId = null;
    _activeRequestId = null;
    _recentPositions.clear();
    _isTracking = false;

    print('📍 DriverLocationTracker: Stopped tracking');
  }

  /// Sample current location and publish if movement is significant.
  Future<void> _sampleLocation() async {
    if (!_isTracking || _activeCellId == null || _activeRequestId == null) {
      return;
    }

    try {
      final position = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.high,
      );

      final shouldPublish = _shouldPublish(position);

      // Add to buffer
      _recentPositions.add(position);
      if (_recentPositions.length > _bufferSize) {
        _recentPositions.removeAt(0);
      }

      if (shouldPublish) {
        await _publishLocation(position);
      } else {
        print('📍 DriverLocationTracker: Skipping publish (minimal movement)');
      }
    } catch (e) {
      print('❌ DriverLocationTracker: Error sampling location: $e');
    }
  }

  /// Determine if position should be published based on movement from buffer average.
  bool _shouldPublish(Position current) {
    // Always publish first position
    if (_recentPositions.isEmpty) {
      return true;
    }

    // Calculate average of recent positions
    double avgLat = 0;
    double avgLng = 0;
    for (final pos in _recentPositions) {
      avgLat += pos.latitude;
      avgLng += pos.longitude;
    }
    avgLat /= _recentPositions.length;
    avgLng /= _recentPositions.length;

    // Calculate distance from average
    final distance = Geolocator.distanceBetween(
      avgLat,
      avgLng,
      current.latitude,
      current.longitude,
    );

    // Only publish if movement exceeds threshold
    return distance > _movementThresholdMeters;
  }

  /// Publish location to Firebase.
  Future<void> _publishLocation(Position position) async {
    if (_activeCellId == null || _activeRequestId == null) return;

    await _requestService.updateDriverLocation(
      _activeCellId!,
      _activeRequestId!,
      position.latitude,
      position.longitude,
    );

    print(
      '📍 DriverLocationTracker: Published location (${position.latitude}, ${position.longitude})',
    );
  }

  /// Dispose of resources.
  void dispose() {
    stopTracking();
  }
}

File: ios/bicitaxi-conductor/bicitaxi-conductor/Services/DriverLocationTracker.swift
//
//  DriverLocationTracker.swift
//  bicitaxi-conductor
//
//  Smart GPS location tracker for conductors.
//  Implements intelligent location publishing to reduce Firebase writes.
//

import Foundation
import CoreLocation
import Combine

/// Smart GPS location tracker for conductors.
///
/// Implements intelligent location publishing to reduce Firebase writes:
/// - Samples GPS every 30 seconds locally
/// - Maintains rolling buffer of last 3 positions
/// - Only publishes to Firebase if current position differs >3m from buffer average
@MainActor
class DriverLocationTracker: NSObject, ObservableObject {
    
    // MARK: - Published Properties
    
    @Published var isTracking: Bool = false
    @Published var lastPublishedLocation: CLLocationCoordinate2D?
    
    // MARK: - Private Properties
    
    private let requestService: RequestService
    private let locationManager = CLLocationManager()
    
    private var locationTimer: Timer?
    private var activeCellId: String?
    private var activeRequestId: String?
    
    // Rolling buffer of recent positions for smart filtering
    private var recentPositions: [CLLocation] = []
    private let bufferSize = 3
    private let movementThresholdMeters: Double = 3.0
    private let samplingInterval: TimeInterval = 30
    
    // MARK: - Initialization
    
    init(requestService: RequestService) {
        self.requestService = requestService
        super.init()
        
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
    }
    
    // MARK: - Public Methods
    
    /// Start tracking driver location for a request.
    /// Call this when driver accepts a request.
    func startTracking(cellId: String, requestId: String) {
        if isTracking {
            stopTracking()
        }
        
        activeCellId = cellId
        activeRequestId = requestId
        recentPositions.removeAll()
        isTracking = true
        
        print("📍 DriverLocationTracker: Started tracking for request \(requestId)")
        
        // Request location permission if needed
        let status = locationManager.authorizationStatus
        if status == .notDetermined {
            locationManager.requestWhenInUseAuthorization()
        }
        
        // Immediately sample first position
        sampleLocation()
        
        // Start periodic sampling
        locationTimer = Timer.scheduledTimer(withTimeInterval: samplingInterval, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.sampleLocation()
            }
        }
    }
    
    /// Stop tracking driver location.
    /// Call this when request is completed or cancelled.
    func stopTracking() {
        locationTimer?.invalidate()
        locationTimer = nil
        activeCellId = nil
        activeRequestId = nil
        recentPositions.removeAll()
        isTracking = false
        lastPublishedLocation = nil
        
        print("📍 DriverLocationTracker: Stopped tracking")
    }
    
    // MARK: - Private Methods
    
    /// Sample current location and publish if movement is significant.
    private func sampleLocation() {
        guard isTracking, activeCellId != nil, activeRequestId != nil else { return }
        
        locationManager.requestLocation()
    }
    
    /// Determine if position should be published based on movement from buffer average.
    private func shouldPublish(current: CLLocation) -> Bool {
        // Always publish first position
        guard !recentPositions.isEmpty else { return true }
        
        // Calculate average of recent positions
        var avgLat: Double = 0
        var avgLng: Double = 0
        for pos in recentPositions {
            avgLat += pos.coordinate.latitude
            avgLng += pos.coordinate.longitude
        }
        avgLat /= Double(recentPositions.count)
        avgLng /= Double(recentPositions.count)
        
        let avgLocation = CLLocation(latitude: avgLat, longitude: avgLng)
        
        // Calculate distance from average
        let distance = current.distance(from: avgLocation)
        
        // Only publish if movement exceeds threshold
        return distance > movementThresholdMeters
    }
    
    /// Publish location to Firebase.
    private func publishLocation(_ location: CLLocation) async {
        guard let cellId = activeCellId, let requestId = activeRequestId else { return }
        
        await requestService.updateDriverLocation(
            cellId: cellId,
            requestId: requestId,
            lat: location.coordinate.latitude,
            lng: location.coordinate.longitude
        )
        
        lastPublishedLocation = location.coordinate
        print("📍 DriverLocationTracker: Published location (\(location.coordinate.latitude), \(location.coordinate.longitude))")
    }
}

// MARK: - CLLocationManagerDelegate

extension DriverLocationTracker: CLLocationManagerDelegate {
    
    nonisolated func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        
        Task { @MainActor in
            let shouldPublishNow = self.shouldPublish(current: location)
            
            // Add to buffer
            self.recentPositions.append(location)
            if self.recentPositions.count > self.bufferSize {
                self.recentPositions.removeFirst()
            }
            
            if shouldPublishNow {
                await self.publishLocation(location)
            } else {
                print("📍 DriverLocationTracker: Skipping publish (minimal movement)")
            }
        }
    }
    
    nonisolated func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print("❌ DriverLocationTracker: Location error: \(error)")
    }
}
